share/p2p/shrexnd/server.go

```diff
@@ -95,90 +141,65 @@ func (srv *Server) handleNamespacedData(ctx context.Context, stream network.Stre
-			srv.respondNotFoundError(ctx, logger, stream)
-			return
+			return nil, pb.StatusCode_NOT_FOUND, nil
 		}
-		logger.Errorw("server: retrieving DAH", "err", err)
-		srv.respondInternalError(ctx, logger, stream)
-		return
+		return nil, pb.StatusCode_INTERNAL, fmt.Errorf("retrieving DAH: %w", err)
 	}
 
-	shares, err := srv.getter.GetSharesByNamespace(ctx, dah, req.Namespace)
-	switch {
-	case errors.Is(err, share.ErrNotFound):
-		logger.Warn("server: nd not found")
-		srv.respondNotFoundError(ctx, logger, stream)
-		return
-	case err != nil:
-		logger.Errorw("server: retrieving shares", "err", err)
-		srv.respondInternalError(ctx, logger, stream)
-		return
-	}
-
-	resp := namespacedSharesToResponse(shares)
-	srv.respond(ctx, logger, stream, resp)
-}
-
-// validateRequest checks correctness of the request
-func validateRequest(req pb.GetSharesByNamespaceRequest) error {
-	if err := share.Namespace(req.Namespace).ValidateForData(); err != nil {
-		return err
+	shares, err := srv.getter.GetSharesByNamespace(ctx, dah, namespace)
+	if err != nil {
+		return nil, pb.StatusCode_INTERNAL, fmt.Errorf("retrieving shares: %w", err)
 	}
-	if len(req.RootHash) != sha256.Size {
-		return fmt.Errorf("incorrect root hash length: %v", len(req.RootHash))
+
+	if shares.IsEmpty() {
+		return shares, pb.StatusCode_NAMESPACE_NOT_FOUND, nil
 	}
-	return nil
+	return shares, pb.StatusCode_OK, nil
 }
 
-// respondNotFoundError sends a not found response to client
-func (srv *Server) respondNotFoundError(ctx context.Context,
-	logger *zap.SugaredLogger, stream network.Stream) {
-	resp := &pb.GetSharesByNamespaceResponse{
-		Status: pb.StatusCode_NOT_FOUND,
+func (srv *Server) respondStatus(
+	logger *zap.SugaredLogger,
+	stream network.Stream,
+	status pb.StatusCode) error {
+	err := stream.SetWriteDeadline(time.Now().Add(srv.params.ServerWriteTimeout))
+	if err != nil {
+		logger.Debugw("server: setting write deadline", "err", err)
 	}
-	srv.respond(ctx, logger, stream, resp)
-}
 
-// respondInternalError sends internal error response to client
-func (srv *Server) respondInternalError(ctx context.Context,
-	logger *zap.SugaredLogger, stream network.Stream) {
-	resp := &pb.GetSharesByNamespaceResponse{
-		Status: pb.StatusCode_INTERNAL,
+	_, err = serde.Write(stream, &pb.GetSharesByNamespaceStatusResponse{Status: status})
+	if err != nil {
+		return fmt.Errorf("writing response: %w", err)
 	}
-	srv.respond(ctx, logger, stream, resp)
+
+	return nil
 }
 
-// namespacedSharesToResponse encodes shares into proto and sends it to client with OK status code
-func namespacedSharesToResponse(shares share.NamespacedShares) *pb.GetSharesByNamespaceResponse {
-	rows := make([]*pb.Row, 0, len(shares))
+// sendNamespacedShares encodes shares into proto messages and sends it to client
+func (srv *Server) sendNamespacedShares(shares share.NamespacedShares, stream network.Stream) error {
 	for _, row := range shares {
-		row := &pb.Row{
+		row := &pb.NamespaceRowResponse{
```

Comment:
Ok, row by row makes total sense. It would be nice to update the ADR with this protocol change, but not for now, as it's not even merged.